src/rgw/rgw_sal_d3n.cc

```diff
@@ -0,0 +1,231 @@
+
+      r = d->client_cb->handle_data(astate->data, obj_ofs, chunk_len);
+      if (r < 0)
+        return r;
+
+      len -= chunk_len;
+      d->offset += chunk_len;
+      read_ofs += chunk_len;
+      obj_ofs += chunk_len;
+      if (!len)
+        return 0;
+    }
+
+    auto obj = d->rgwrados->svc.rados->obj(read_obj);
+    r = obj.open(dpp);
+    if (r < 0) {
+      ldpp_dout(dpp, 20) << "failed to open rados context for " << read_obj << dendl;
+      return r;
+    }
+
+    ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb::" << __func__ << "(): oid=" << read_obj.oid << " obj-ofs=" << obj_ofs << " read_ofs=" << read_ofs << " len=" << len << dendl;
+    op.read(read_ofs, len, nullptr, nullptr);
+
+    const uint64_t cost = len;
+    const uint64_t id = obj_ofs; // use logical object offset for sorting replies
+
+    auto completed = d->aio->get(obj, rgw::Aio::librados_op(std::move(op), d->yield), cost, id);
+    return d->flush(std::move(completed));
+  } else {
+    ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb::" << __func__ << "(): oid=" << read_obj.oid << ", is_head_obj=" << is_head_obj << ", obj-ofs=" << obj_ofs << ", read_ofs=" << read_ofs << ", len=" << len << dendl;
+    int r;
+
+    op.read(read_ofs, len, nullptr, nullptr);
+
+    const uint64_t cost = len;
+    const uint64_t id = obj_ofs; // use logical object offset for sorting replies
+    oid = read_obj.oid;
+
+    auto obj = d->rgwrados->svc.rados->obj(read_obj);
+    r = obj.open(dpp);
+    if (r < 0) {
+      ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb:: Error: failed to open rados context for " << read_obj << ", r=" << r << dendl;
+      return r;
+    }
+
+    const bool is_compressed = (astate->attrset.find(RGW_ATTR_COMPRESSION) != astate->attrset.end());
+    const bool is_encrypted = (astate->attrset.find(RGW_ATTR_CRYPT_MODE) != astate->attrset.end());
+    if (read_ofs != 0 || astate->size != astate->accounted_size || is_compressed || is_encrypted) {
+      d->d3n_bypass_cache_write = true;
+      ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb:: " << __func__ << "(): Note - bypassing datacache: oid=" << read_obj.oid << ", read_ofs!=0 = " << read_ofs << ", size=" << astate->size << " != accounted_size=" << astate->accounted_size << ", is_compressed=" << is_compressed << ", is_encrypted=" << is_encrypted  << dendl;
+      auto completed = d->aio->get(obj, rgw::Aio::librados_op(std::move(op), d->yield), cost, id);
+      r = d->flush(std::move(completed));
+      return r;
+    }
+
+    if (filter->get_d3n_cache()->get(oid, len)) {
+      // Read From Cache
+      ldpp_dout(dpp, 20) << "D3NFilterObject::get_obj_iterate_cb:: " << __func__ << "(): READ FROM CACHE: oid=" << read_obj.oid << ", obj-ofs=" << obj_ofs << ", read_ofs=" << read_ofs << ", len=" << len << dendl;
+      auto completed = d->aio->get(obj, rgw::Aio::d3n_cache_op(dpp, d->yield, read_ofs, len, d->rgwrados->d3n_data_cache->cache_location), cost, id);
```

Comment:
there is one thing that I missed doing - removing all references to d3n_data_cache from rados store, which is why this ->d->rgwrados->d3n_data_cache->cache_location, still works fine, and the flush() method in rados also internally uses the reference to d3n_data_cache. I will remove these references while working towards making it generic for all backend stores.